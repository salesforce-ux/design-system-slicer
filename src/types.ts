import Immutable from 'immutable';

export type Selector = string;

export type Cache = CacheItem[];

export interface CacheItem {
  selectors: string[];
  relatedSelectors?: string[];
  type: string;
  css: string;
}

export type CacheBuildState = Immutable.Map<string, Cache>;
