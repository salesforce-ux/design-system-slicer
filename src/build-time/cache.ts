// Copyright (c) 2015-present, salesforce.com, inc. All rights reserved
// Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license

import Immutable from 'immutable';

import {
  Reducer,
  extractTags,
  extractClassNames,
  extractSelectorsFromAtRuleAndRecurse
} from './reducers';

import { Cache, CacheItem } from '../types';

import { getAllRulesFromCss, Rule } from './lib/parse-css';

const extractTagsAndClassNames: Reducer = extractTags.concat(extractClassNames);

// order matters
const extractSelectorPartsAndAssociateCss: Reducer = extractSelectorsFromAtRuleAndRecurse(
  extractTagsAndClassNames
).concat(extractTagsAndClassNames);

const mapSelectorsToMatchingCss = (rules: Rule[]): Cache =>
  Immutable.List(rules).reduce(
    extractSelectorPartsAndAssociateCss.run,
    Immutable.List()
  );

export const create = (css: string): Promise<Cache> =>
  getAllRulesFromCss(css).then(mapSelectorsToMatchingCss);
