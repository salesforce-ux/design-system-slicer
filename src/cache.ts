// Copyright (c) 2015-present, salesforce.com, inc. All rights reserved
// Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license

import Immutable from 'immutable';

import postcss, {
  Root as PostCssRoot,
  Rule as PostCssRule,
  AtRule as PostCssAtRule,
  ChildNode as PostCssChildNode,
  Result as PostCssResult
} from 'postcss';

import {
  Reducer,
  htmlReducer,
  classNameReducer,
  atRuleReducer,
  miscReducer
} from './reducers';

import {
  ComponentName,
  ComponentLookup,
  Selector,
  RootSelectors,
  CacheBuildState,
  Cache,
  AtRuleAdapter,
  Rule
} from './types';

import _ from 'lodash';

function atRuleToRule(atrule: PostCssAtRule): AtRuleAdapter {
  return { selector: '', type: 'atrule', rule: atrule };
}

function handleRule(
  visitor: (rule: Rule) => void,
  node: PostCssChildNode
): void {
  switch (node.type) {
    case 'comment':
      node.remove();
      break;
    case 'rule':
      visitor(node);
      break;
    case 'atrule':
      visitor(atRuleToRule(node));
      break;
  }
}

function visitorPlugin(visitor: (rule: Rule) => void) {
  return (root: PostCssRoot, result?: PostCssResult): void => {
    root.walk(r => handleRule(visitor, r));
  };
}

function walkCss(css: string, visitor: (rule: Rule) => void): Promise<string> {
  return postcss()
    .use(visitorPlugin(visitor))
    .process(css, { from: undefined })
    .then(result => result.css);
}

function allRules(css: string): Promise<Immutable.List<Rule>> {
  const acc: Rule[] = [];
  return walkCss(css, rule => acc.push(rule)).then(() => Immutable.List(acc));
}

// order matters
const reduction: Reducer = atRuleReducer
  .concat(htmlReducer)
  .concat(classNameReducer)
  .concat(miscReducer);

function build(rules: Immutable.List<Rule>): Cache {
  return rules.reduce(reduction.run, Immutable.List());
}
export function create(
  rootSelectors: RootSelectors,
  utilities: string[],
  css: string
): Promise<Cache> {
  return allRules(css).then(build);
}
