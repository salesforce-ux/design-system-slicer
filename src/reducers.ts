import Immutable from 'immutable';

import postcss, { Rule as PostCssRule } from 'postcss';

import { ComponentName, ComponentLookup, Selector } from './types';
import escapeRegExp from 'lodash.escaperegexp';

export type ReducingFn<S = any> = (state: S, value: any) => S;

export type Semigroup<S = any> = (x: S) => S;

export type Reducer = { run: ReducingFn; concat: Semigroup<Reducer> };

const reducer = (run: ReducingFn): Reducer => ({
  run,
  concat: other => reducer((acc, rule) => other.run(run(acc, rule), rule))
});

function fromNullable<T>(value?: T): Immutable.List<T> {
  return value ? Immutable.List.of(value) : Immutable.List();
}

function findRootSelector(
  rootSelectors: Immutable.List<string>,
  selector: string
): Immutable.List<string> {
  return fromNullable(
    rootSelectors.find(rootSelector =>
      selector
        .split(',')
        .some(
          part =>
            !!part.match(
              new RegExp(`^(\\w+)?${escapeRegExp(rootSelector)}`, 'ig')
            )
        )
    )
  );
}

const setFirstSelectorReducer: Reducer = reducer(
  (acc, rule) =>
    acc.get('firstSelectorFound')
      ? acc
      : acc.set('firstSelectorFound', /^\./.test(rule.selector))
);

const normalizeReducer: Reducer = reducer(
  (acc, rule) =>
    !acc.get('firstSelectorFound')
      ? acc.updateIn(
          ['result', 'normalize'],
          (maybeRules: Immutable.List<PostCssRule> | undefined) => {
            let rules = maybeRules
              ? (maybeRules as Immutable.List<PostCssRule>)
              : Immutable.List();
            return rules.push(rule);
          }
        )
      : acc
);

// We only support [class*=] right now
function selectorFromComplex(selector: Selector): String | undefined {
  const match = selector.match(/\[class\*='([a-zA-Z\-\_]+)'\]/);
  return match != null ? match[1] : undefined;
}

function isComplexSelector(selector: Selector): Boolean {
  return !!selectorFromComplex(selector);
}

function insertRuleToCache(
  acc: Immutable.OrderedMap<ComponentName, Immutable.List<PostCssRule>>,
  selector: Selector | undefined,
  rule: PostCssRule
): Immutable.OrderedMap<ComponentName, Immutable.List<PostCssRule>> {
  return selector
    ? acc.updateIn(
        ['result', selector],
        (maybeRules: Immutable.List<PostCssRule> | undefined) => {
          let rules = maybeRules
            ? (maybeRules as Immutable.List<PostCssRule>)
            : Immutable.List();
          return rules.push(rule);
        }
      )
    : acc;
}

function utilReducer(utilities: Immutable.Set<Selector>): Reducer {
  return reducer((acc, rule) => {
    const foundUtil: Selector | undefined = utilities.find(
      u =>
        isComplexSelector(u)
          ? rule.selector.match(selectorFromComplex(u))
          : rule.selector.includes(u)
    );
    return insertRuleToCache(acc, foundUtil, rule);
  });
}

function componentReducer(
  rootSelectors: Immutable.List<Selector>,
  lookup: ComponentLookup
): Reducer {
  return reducer(
    (acc, rule) =>
      findRootSelector(rootSelectors, rule.selector)
        .map(key => lookup.get(key))
        .map(foundComponent => insertRuleToCache(acc, foundComponent, rule))
        .first() || acc
  );
}

export {
  setFirstSelectorReducer,
  normalizeReducer,
  utilReducer,
  componentReducer
};
