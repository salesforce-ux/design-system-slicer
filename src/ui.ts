// Copyright (c) 2015-present, salesforce.com, inc. All rights reserved
// Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license

export interface UI {
  components: {
    [propName: string]: UINode;
  };
  utilities: {
    [propName: string]: UINode;
  };
}

export interface UINode {
  annotations: UINodeAnnotations;
  restrictees: UINode[];
}

export interface UINodeAnnotations {
  selector: string;
  restrict: string;
}

export function reduceNode<T>(
  reducer: (result: T, node: UINode) => T,
  empty: T,
  node: UINode
): T {
  return node.restrictees.reduce(
    (result, n) => reduceNode(reducer, result, n),
    reducer(empty, node)
  );
}

export function reduceNodeRight<T>(
  reducer: (result: T, node: UINode) => T,
  empty: T,
  node: UINode
): T {
  return reducer(
    node.restrictees.reduceRight(
      (result, n) => reduceNodeRight(reducer, result, n),
      empty
    ),
    node
  );
}

export function toList(node: UINode): UINode[] {
  return reduceNodeRight<UINode[]>((a, b) => a.concat(b), [], node);
}
