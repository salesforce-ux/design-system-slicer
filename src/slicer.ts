// Copyright (c) 2015-present, salesforce.com, inc. All rights reserved
// Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license

import { Cache } from './cache';

export type Css = string;

export function sliceForComponents(cache: Cache, ...components: string[]): Css {
  return cache
    .filter(slice => components.includes(slice.name))
    .reduce(
      (css, slice) => css + slice.rules.map(r => r.toString()).join('\n'),
      ''
    );
}

export class Slicer {
  private cache: Cache;
  constructor(cache: Cache) {
    this.cache = cache;
  }
  sliceForComponents(...components: string[]): Css {
    return sliceForComponents(this.cache, ...components);
  }
}

export function create(cache: Cache): Slicer {
  return new Slicer(cache);
}
