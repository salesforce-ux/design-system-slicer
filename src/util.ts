// Copyright (c) 2015-present, salesforce.com, inc. All rights reserved
// Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license

import { UI, UINode, toList } from './ui';

export function allSelectorsForComponent(node: UINode): string[] {
  return node.annotations.selector.split(',').map(s => s.trim());
}

export function findBySelector(node: UINode, selector: string): UINode[] {
  return toList(node).filter(n => n.annotations.selector === selector);
}

export class Util {
  private ui: UI;
  private _utils: Array<UINode> | undefined;
  constructor(ui: UI) {
    this.ui = ui;
  }
  utilities(): Array<UINode> {
    return (
      this._utils ||
      (this._utils = Object.keys(this.ui.utilities)
        .map(key => this.ui.utilities[key])
        .reduce<UINode[]>((result, node) => result.concat(toList(node)), []))
    );
  }
  components(): Array<UINode> {
    return Object.keys(this.ui.components).map(key => this.ui.components[key]);
  }
  componentsForSelector(selector: string): string[] {
    return Object.keys(this.ui.components).filter(
      key => findBySelector(this.ui.components[key], selector).length > 0
    );
  }
  rootSelectors(componentName: string): string[] {
    let component = this.ui.components[componentName];
    let selectors = component.annotations.selector
      .split(',')
      .map(s => s.trim());
    return component.restrictees
      .map(node =>
        node.restrictees
          .map(
            node =>
              selectors.some(rootSelector =>
                node.annotations.restrict.endsWith(rootSelector)
              )
                ? node.annotations.selector
                : ''
          )
          .filter(x => x)
      )
      .reduce((a, b) => a.concat(b), selectors);
  }
  isUtility(selector: string): boolean {
    return !!this.utilities().filter(
      n => n.annotations.selector === selector
    )[0];
  }
}

export function create(ui: UI): Util {
  return new Util(ui);
}
