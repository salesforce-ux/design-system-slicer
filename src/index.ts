// Copyright (c) 2015-present, salesforce.com, inc. All rights reserved
// Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license

// @ts-ignore
import cache from '../dist/cache.json';
// @ts-ignore
import ui from '@salesforce-ux/design-system/ui.json';

import { Cache } from './types';
import { Slicer, create as createSlicer } from './run-time/slicer';
import { Util, create as createUtil } from './run-time/util';

export { createSlicer, createUtil };

export const slicer = createSlicer(cache as Cache);
export const util = createUtil(ui);
