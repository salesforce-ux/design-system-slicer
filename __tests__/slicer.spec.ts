// Copyright (c) 2015-present, salesforce.com, inc. All rights reserved
// Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license

import * as fs from 'fs';
import * as path from 'path';

import * as Slicer from '../src';
import { create as createCache, Cache } from '../src/cache';

// ui.json
// -----------
// components: datepicker, buttons, modal
// utilities: interactions, hyphenation, alignment

// css format
// -----------
// global resets
// datepicker
// modal
// button-group
// button
// utils (just hyphenation though)
//

const css = fs
  .readFileSync(path.resolve(__dirname, '__fixtures__/slds.css'))
  .toString();

let cache: Cache;

beforeEach(async () => {
  cache = await createCache(
    {
      buttons: ['.slds-button', 'slds-button_neutral'],
      datepickers: ['.slds-datepicker']
    },
    css
  );
});

it('buttons slice', () => {
  const slicer = Slicer.create(cache);
  const result = slicer.getSliceForComponents('buttons');

  expect(result).toMatch('.slds-button');
  expect(result).not.toMatch('.slds-datepicker');
  expect(result).not.toMatch('.slds-truncate_container_66');
});

it('date-picker slice', () => {
  const slicer = Slicer.create(cache);
  const result = slicer.getSliceForComponents('datepickers');

  expect(result).toMatch('.slds-datepicker');
  expect(result).not.toMatch('.slds-button');
  expect(result).not.toMatch('.slds-truncate_container_66');
});
