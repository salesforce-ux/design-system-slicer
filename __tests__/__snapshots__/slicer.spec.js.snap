// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button css matches the snapshot 1`] = `
".slds-button-group,
.slds-button-group-list {
  display: -ms-inline-flexbox;
  display: inline-flex; }

.slds-button-group .slds-button,
.slds-button-group-list .slds-button {
  border-radius: 0;
  border-width: 1px; }

.slds-button-group .slds-button:focus,
.slds-button-group-list .slds-button:focus {
  z-index: 1; }

.slds-button-group .slds-button + .slds-button,
.slds-button-group .slds-button + .slds-button_last .slds-button,
.slds-button-group .slds-button + .slds-button--last .slds-button,
.slds-button-group-list li + li .slds-button {
  margin-left: -1px; }

.slds-button-group-list li:first-child .slds-button,
.slds-button-group .slds-button:first-child {
  border-radius: 0.25rem 0 0 0.25rem; }
.slds-button {
  position: relative;
  display: inline-block;
  padding: 0;
  background: transparent;
  background-clip: border-box;
  border: 1px solid transparent;
  border-radius: 0.25rem;
  line-height: 1.875rem;
  text-decoration: none;
  color: #0070d2;
  -webkit-appearance: none;
  white-space: normal;
  -webkit-user-select: none;
     -moz-user-select: none;
      -ms-user-select: none;
          user-select: none; }

.slds-button:hover, .slds-button:focus, .slds-button:active, .slds-button:visited {
  text-decoration: none; }

.slds-button:hover, .slds-button:focus {
  color: #005fb2; }

.slds-button:focus {
  outline: 0;
  box-shadow: 0 0 3px #0070D2; }

.slds-button:active {
  color: #005fb2; }

.slds-button[disabled] {
  color: #dddbda; }

.slds-button[disabled] * {
  pointer-events: none; }

.slds-button a {
  color: currentColor; }

.slds-button:hover .slds-button__icon, .slds-button:focus .slds-button__icon, .slds-button:active .slds-button__icon, .slds-button[disabled] .slds-button__icon {
  fill: currentColor;
  pointer-events: none; }

.slds-button + .slds-button-group,
.slds-button + .slds-button-group-list {
  margin-left: 0.25rem; }

.slds-button + .slds-button {
  margin-left: 0.25rem; }
.slds-button-space-left {
  margin-left: 0.25rem; }

"
`;

exports[`button, datepicker css matches the snapshot 1`] = `
".slds-datepicker {
  padding: 0;
  font-size: 0.75rem; }

.slds-datepicker th,
.slds-datepicker td {
  text-align: center; }

.slds-datepicker th {
  padding: 0.5rem;
  font-weight: 400;
  color: #706e6b; }

.slds-datepicker td {
  padding: 0.25rem;
  text-align: center;
  font-size: 0.75rem; }

.slds-datepicker td > .slds-day {
  width: 2rem;
  height: 2rem;
  display: block;
  position: relative;
  min-width: 2rem;
  line-height: 2rem;
  border-radius: 50%;
  margin: auto; }

.slds-datepicker td:hover:not(.slds-disabled-text) > .slds-day,
.slds-datepicker td:focus:not(.slds-disabled-text) > .slds-day,
.slds-datepicker td.slds-is-today > .slds-day {
  cursor: pointer; }

.slds-datepicker td.slds-is-today > .slds-day {
  background-color: #ecebea; }

.slds-datepicker td:focus {
  outline: 0; }

.slds-datepicker td:focus > .slds-day {
  box-shadow: #0070d2 0 0 0 1px inset; }

.slds-datepicker td.slds-is-selected > .slds-day {
  background: #0070d2;
  color: white; }

.slds-datepicker td.slds-is-selected:focus > .slds-day {
  background: #005fb2;
  box-shadow: #005fb2 0 0 3px;
  color: white; }

.slds-datepicker { }

.slds-datepicker .slds-has-multi-selection .slds-is-selected-multi:last-child > .slds-day:after {
  left: auto;
  right: 0;
  transform: translateX(0.25rem); }

.slds-datepicker .slds-has-multi-row-selection .slds-is-selected-multi:first-child > .slds-day:before,
.slds-datepicker .slds-has-multi-row-selection .slds-is-selected-multi:last-child > .slds-day:after {
  content: '';
  position: absolute;
  background: #0070d2;
  top: 0;
  left: -50%;
  height: 100%;
  width: 2.5rem;
  transform: translateX(-0.5rem);
  z-index: -1; }

.slds-datepicker .slds-has-multi-row-selection .slds-is-selected-multi:first-child > .slds-day:before {
  left: 0;
  transform: translateX(-0.25rem); }

.slds-datepicker .slds-has-multi-row-selection .slds-is-selected-multi:last-child > .slds-day:after {
  left: auto;
  right: 0;
  transform: translateX(0.25rem); }

.slds-datepicker td.slds-is-selected-multi > .slds-day {
  overflow: visible; }

.slds-datepicker td.slds-is-selected-multi.slds-is-today > .slds-day {
  box-shadow: white 0 0 0 1px inset; }

.slds-datepicker td.slds-is-selected-multi + .slds-is-selected-multi > .slds-day:before {
  content: '';
  position: absolute;
  background: #0070d2;
  top: 0;
  left: -50%;
  height: 100%;
  width: 2.5rem;
  transform: translateX(-0.5rem);
  z-index: -1; }
.slds-button-group,
.slds-button-group-list {
  display: -ms-inline-flexbox;
  display: inline-flex; }

.slds-button-group .slds-button,
.slds-button-group-list .slds-button {
  border-radius: 0;
  border-width: 1px; }

.slds-button-group .slds-button:focus,
.slds-button-group-list .slds-button:focus {
  z-index: 1; }

.slds-button-group .slds-button + .slds-button,
.slds-button-group .slds-button + .slds-button_last .slds-button,
.slds-button-group .slds-button + .slds-button--last .slds-button,
.slds-button-group-list li + li .slds-button {
  margin-left: -1px; }

.slds-button-group-list li:first-child .slds-button,
.slds-button-group .slds-button:first-child {
  border-radius: 0.25rem 0 0 0.25rem; }
.slds-button {
  position: relative;
  display: inline-block;
  padding: 0;
  background: transparent;
  background-clip: border-box;
  border: 1px solid transparent;
  border-radius: 0.25rem;
  line-height: 1.875rem;
  text-decoration: none;
  color: #0070d2;
  -webkit-appearance: none;
  white-space: normal;
  -webkit-user-select: none;
     -moz-user-select: none;
      -ms-user-select: none;
          user-select: none; }

.slds-button:hover, .slds-button:focus, .slds-button:active, .slds-button:visited {
  text-decoration: none; }

.slds-button:hover, .slds-button:focus {
  color: #005fb2; }

.slds-button:focus {
  outline: 0;
  box-shadow: 0 0 3px #0070D2; }

.slds-button:active {
  color: #005fb2; }

.slds-button[disabled] {
  color: #dddbda; }

.slds-button[disabled] * {
  pointer-events: none; }

.slds-button a {
  color: currentColor; }

.slds-button:hover .slds-button__icon, .slds-button:focus .slds-button__icon, .slds-button:active .slds-button__icon, .slds-button[disabled] .slds-button__icon {
  fill: currentColor;
  pointer-events: none; }

.slds-button + .slds-button-group,
.slds-button + .slds-button-group-list {
  margin-left: 0.25rem; }

.slds-button + .slds-button {
  margin-left: 0.25rem; }
.slds-button-space-left {
  margin-left: 0.25rem; }

"
`;

exports[`normalize matches the snapshot 1`] = `
"html {
  font-family: sans-serif;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%; }

body {
  margin: 0; }

article,
aside,
details,
figcaption,
figure,
footer,
header,
hgroup,
main,
menu,
nav,
section,
summary {
  display: block; }

audio,
canvas,
progress,
video {
  display: inline-block;
  vertical-align: baseline; }

audio:not([controls]) {
  display: none;
  height: 0; }

template {
  display: none; }

a {
  background-color: transparent; }

a:active,
a:hover {
  outline: 0; }

abbr[title] {
  border-bottom: 1px dotted; }

b,
strong {
  font-weight: bold; }

dfn {
  font-style: italic; }

h1 {
  font-size: 2em;
  margin: 0.67em 0; }

mark {
  background: #ff0;
  color: #000; }

small {
  font-size: 80%; }

sub,
sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline; }

sup {
  top: -0.5em; }

sub {
  bottom: -0.25em; }

img {
  border: 0; }

svg:not(:root) {
  overflow: hidden; }

figure {
  margin: 1em 40px; }

hr {
  -moz-box-sizing: content-box;
  box-sizing: content-box;
  height: 0; }

pre {
  overflow: auto; }

code,
kbd,
pre,
samp {
  font-family: monospace, monospace;
  font-size: 1em; }

button,
input,
optgroup,
select,
textarea {
  color: inherit;
  font: inherit;
  margin: 0; }

button {
  overflow: visible; }

button,
select {
  text-transform: none; }

button,
html input[type=\\"button\\"],
input[type=\\"reset\\"],
input[type=\\"submit\\"] {
  -webkit-appearance: button;
  cursor: pointer; }

button[disabled],
html input[disabled] {
  cursor: default; }

button::-moz-focus-inner,
input::-moz-focus-inner {
  border: 0;
  padding: 0; }

input {
  line-height: normal; }

input[type=\\"checkbox\\"],
input[type=\\"radio\\"] {
  box-sizing: border-box;
  padding: 0; }

input[type=\\"number\\"]::-webkit-inner-spin-button,
input[type=\\"number\\"]::-webkit-outer-spin-button {
  height: auto; }

input[type=\\"search\\"] {
  -webkit-appearance: textfield;
  -moz-box-sizing: content-box;
  -webkit-box-sizing: content-box;
  box-sizing: content-box; }

input[type=\\"search\\"]::-webkit-search-cancel-button,
input[type=\\"search\\"]::-webkit-search-decoration {
  -webkit-appearance: none; }

fieldset {
  border: 1px solid #c0c0c0;
  margin: 0 2px;
  padding: 0.35em 0.625em 0.75em; }

legend {
  border: 0;
  padding: 0; }

textarea {
  overflow: auto; }

optgroup {
  font-weight: bold; }

table {
  border-collapse: collapse;
  border-spacing: 0; }

td,
th {
  padding: 0; }

*,
*:before,
*:after {
  box-sizing: border-box; }

:-ms-input-placeholder {
  color: #706e6b;
  font-weight: 400;
  font-size: 0.8125rem; }

::placeholder {
  color: #706e6b;
  font-weight: 400;
  font-size: 0.8125rem; }

::-moz-selection {
  background: #d8edff;
  text-shadow: none;
  color: #3e3e3c; }

::selection {
  background: #d8edff;
  text-shadow: none;
  color: #3e3e3c; }

html {
  font-family: \\"Salesforce Sans\\", Arial, sans-serif;
  font-size: 100%;
  line-height: 1.5;
  background: #fafaf9;
  color: #3e3e3c;
  -webkit-tap-highlight-color: transparent; }

body {
  font-size: 0.8125rem;
  background: transparent; }
.slds-datepicker {
  padding: 0;
  font-size: 0.75rem; }

"
`;

exports[`utilities matches the snapshot 1`] = `
".slds-truncate_container_66,
.slds-truncate_container--66 {
  max-width: 66%; }
.slds-hyphenate {
  overflow-wrap: break-word;
  word-wrap: break-word;
  -webkit-hyphens: auto;
      -ms-hyphens: auto;
          hyphens: auto; }

"
`;
